### ISSUE TITLE
Table cells widths are uneven.

### ISSUE THREAD

**Issue by ehmicky** [all-contributors/all-contributors-cli#173]
Problem description: on small screens table cells have uneven widths which looks odd.

![Screenshot from 2019-03-11 11-37-06](https://user-images.githubusercontent.com/8136211/54117863-0bc33580-43f2-11e9-8ec3-d9bf452e0b62.png)

This seems to be related to the length of the username and word wrapping.

Reproduction repository: https://github.com/ehmicky/portable-node-guide

Must be looked at on a smartphone.

- `all-contributors-cli` version: `6.1.2`
- `node` version: `11.11.0`
- `npm` (or `yarn`) version: `6.9.0`

Relevant code or config

```javascript
{
  "projectName": "portable-node-guide",
  "projectOwner": "ehmicky",
  "repoType": "github",
  "repoHost": "https://github.com",
  "files": [
    "README.md"
  ],
  "imageSize": 100,
  "commit": true,
  "contributors": [
    {
      "login": "ehmicky",
      "name": "ehmicky",
      "avatar_url": "https://avatars2.githubusercontent.com/u/8136211?v=4",
      "profile": "https://twitter.com/ehmicky",
      "contributions": [
        "code",
        "design",
        "ideas",
        "doc"
      ]
    },
    {
      "login": "thatalextaylor",
      "name": "thatalextaylor",
      "avatar_url": "https://avatars3.githubusercontent.com/u/1481643?v=4",
      "profile": "https://github.com/thatalextaylor",
      "contributions": [
        "ideas",
        "doc"
      ]
    },
    {
      "login": "bnoordhuis",
      "name": "Ben Noordhuis",
      "avatar_url": "https://avatars0.githubusercontent.com/u/275871?v=4",
      "profile": "https://github.com/bnoordhuis",
      "contributions": [
        "ideas",
        "doc"
      ]
    },
    {
      "login": "SteveALee",
      "name": "Steve Lee",
      "avatar_url": "https://avatars2.githubusercontent.com/u/618922?v=4",
      "profile": "http://opendirective.com",
      "contributions": [
        "ideas",
        "doc",
        "talk"
      ]
    },
    {
      "login": "tracker1",
      "name": "Michael J. Ryan",
      "avatar_url": "https://avatars3.githubusercontent.com/u/444316?v=4",
      "profile": "http://tracker1.info/",
      "contributions": [
        "ideas"
      ]
    }
  ],
  "contributorsPerLine": 7
}
```

**Comment by Berkmann18** [comment]
Thanks for reporting this.
I tried a [few things](https://github.com/Berkmann18/portable-node-guide) but since GH blacklists (inline) CSS it's really tricky to get an HTML-only working approach.

@tbenning @jakebolam Would you guys know another solution?

**Comment by jakebolam** [comment]
Thanks for reporting this @ehmicky. Any thoughts on approaches to address this?

**Comment by ehmicky** [comment]
Unfortunately not, sorry :/

I've also noticed that the table is horizontally scrolled on mobile, but I'm not sure if this is intentional and/or this is related.

**Comment by Berkmann18** [comment]
@ehmicky > I've also noticed that the table is horizontally scrolled on mobile, but I'm not sure if this is intentional and/or this is related.

This is definitely not intentional, I'll try fixing that.

**Comment by ehmicky** [comment]
@Berkmann18 you can reproduce by checking the same repository I posted above on a mobile screen. Please let me know if you would like me to post a screenshot.

**Comment by Berkmann18** [comment]
@ehmicky Yes I know, I tried something out but since there's more entries on the row than what my (and certainly your) phone can display, there's that horizontal scrolling even when the table's width is set to `99%`.

**Comment by Berkmann18** [comment]
@ehmicky Have you tried rolling back to the `v5` version where markdown was used for the rendering?

**Comment by ehmicky** [comment]
No I haven't because while this issue leads to horizontal scrolling, I don't think it justifies rolling back (and losing on bug fixes and features).

**Comment by Berkmann18** [comment]
@ehmicky Well, the `v5` (if I recall correctly) was generating Markdown tables instead of HTML ones so it probably will fix the issue or part of it.

If the horizontal scrolling were to re-occur, then either changing the `contributorsPerLine` property in `.all-contributorsrc` or making the table responsive (which might be hard or perhaps impossible given the restrictions on GH file rendering).

**Comment by ehmicky** [comment]
The horizontal scrolling is still an issue (I just tried it), so it will re-occur. 

I tried decreasing `contributorsPerLine`. It looks better on mobile. However it looks worse on desktop.

Locking the version back to a previous major version for the foreseeable future does not look like a good practice, as I might miss important bug fixes, features or security patches.

The real fix would be to make the table responsive, but I'm not sure if this possible :/

**Comment by Berkmann18** [comment]
@ehmicky
> Thanks for trying to figure a solution to this issue :)

> The horizontal scrolling is still an issue (I just tried it), so it will re-occur.

😞.

> I tried decreasing contributorsPerLine. It looks better on mobile. However it looks worse on desktop.

That's what I thought would happen in case of no responsiveness.

> Locking the version back to a previous major version for the foreseeable future does not look like a good practice, as I might miss important bug fixes, features or security patches.

Good point, I was thinking of having either a separate development channel which would be as up-to-date as the upstream but using `v5`'s Markdown rendering approach or having a submodule/subpackage which was generating the table in Markdown as opposed to doing it in HTML?

What are your thoughts about that?

> The real fix would be to make the table responsive, but I'm not sure if this possible :-/

Indeed, however Github's renderer strips out `<script>`, `<style>`, `style="..."` which makes this sort of thing hard (if not impossible).

**Comment by ehmicky** [comment]
> Good point, I was thinking of having either a separate development channel which would be as up-to-date as the upstream but using v5's Markdown rendering approach or having a submodule/subpackage which was generating the table in Markdown as opposed to doing it in HTML?
> What are your thoughts about that?

Wouldn't it be simpler to have both options in the same branch in v6 but being a boolean option?

**Comment by Berkmann18** [comment]
@ehmicky Yeah, that's a better idea.
@sinchang @jakebolam What do you guys think?

**Comment by qyurila** [comment]
After some trials and errors, I ended up reaching a conclusion that specifing `width` attribute of `<td>` tag would be a best solution.

## [Preview branch](https://github.com/qyurila/cli/tree/temp/fixed-width-td-prediew)

Personally, it looks best that the combination of fixed width `<td>` and `contributorsPerLine: 6` config option (2nd table of above preview), in context of both of repo view and [file view](https://github.com/all-contributors/all-contributors/blob/master/README.md). The one with `contributorsPerLine: 7` does not guarantee given avatar width (100px) in repo view. The tag would be like this:

```html
<td align="center" width="16.66%">
  <a href="#">
    <img src="#" width="100" alt="Foo Bar"/><br />
    <sub><b>Foo Bar</b></sub>
  </a>
  <br /><a href="#" title="Code">💻</a>
</td>
```

It requires manual calculation in order to get proper percentage value (because we cannot use `calc()` here), but implementing it shouldn't be too hard.

I tested markdown table too, but it shown no visual difference between current implement (not fixed width).

**Comment by qyurila** [comment]
Additionally, setting `valign` attribute of `<td>` to `"top"` might result the better visual. I added the example about it to preview repo, too.

**Comment by tenshiAMD** [comment]
@qyurila hi, thanks for suggesting these workarounds. You wanna give this a shot? PRs are very welcome.

- using `width="16.66%"` (we can compute this percentage programmatically which is`contributorsPerLine` divided by `contributorsPerLine`)
- using `valign` (unfortunately this is already deprecated  in HTML5)

**Comment by Berkmann18** [comment]
@qyurila @ehmicky Can either of you try with v6.24 and see if it's still an issue?

**Comment by qyurila** [comment]
> @qyurila @ehmicky Can either of you try with v6.24 and see if it's still an issue?

@Berkmann18 The visual got somewhat better but the issue is not quite solved.

You can check [this branch of my repo](https://github.com/qyurila/all-contributors-cli/tree/temp/v6.24.0) to see what the current generation looks like.

**Comment by Berkmann18** [comment]
@qyurila Do you have a solution or suggestion for this? I've tried an approach similar to how I did for HTML emails but GH isn't nice when it comes to (inline) CSS so it's tricky to get the table to look clean.

**Comment by qyurila** [comment]
@Berkmann18 I couldn't find any solution to improve this. The main (if not only) cause of the issue is that any long single-word username won't be wrapped, and GitHub simply does not provide any way to force-wrap them. The exception was their mobile app, which always had been showed tables with 100% width IIRC, but some time ago they fixed it.

**Comment by Berkmann18** [comment]
@qyurila That's unfortunate and yeah the limitations GH puts on HTML/CSS makes stuff like this hard to resolve and they seem fairly slow when it comes to responding to suggestions.
